charts/kube-state/templates/kube-state-role.yaml

```diff
@@ -1,18 +1,34 @@
 ################################
 ## Kube State Role/ClusterRole
 #################################
-{{- $singleNamespace := .Values.global.singleNamespace }}
-{{- if .Values.global.rbacEnabled }}
+{{- $useRoles := or .Values.global.features.namespacePools.enabled .Values.global.singleNamespace }}
+{{- $useClusterRoles := and .Values.global.rbacEnabled .Values.global.clusterRoles (not $useRoles )}}
+{{- $shouldCreateResources := and .Values.global.rbacEnabled (or .Values.global.clusterRoles .Values.global.features.namespacePools.enabled) }}
+{{- $namespaces := .Values.global.features.namespacePools.namespaces.names }}
+{{- if $useClusterRoles }}
+  {{- $namespaces = list .Release.Namespace}}
+{{- else  }}
+  {{- $namespaces = append $namespaces .Release.Namespace }}
+{{- end }}
+
+{{- $rootContext := . }}
+
+{{- if $shouldCreateResources }}
+{{- range $i, $namespaceName := $namespaces -}}
+---
 apiVersion: {{ template "apiVersion.rbac" . }}
-kind: {{ if $singleNamespace }}Role{{ else }}ClusterRole{{ end }}
+kind: {{ if $useClusterRoles }}ClusterRole{{ else }}Role{{ end }}
 metadata:
-  name: {{ template "kube-state.fullname" . }}
+  name: {{ template "kube-state.fullname" $rootContext }}
+  {{- if not $useClusterRoles }}
+  namespace: {{ $namespaceName }}
+  {{- end }}
   labels:
     tier: monitoring
-    component: {{ template "kube-state.name" . }}
-    chart: {{ template "kube-state.chart" . }}
-    release: {{ .Release.Name }}
-    heritage: {{ .Release.Service }}
+    component: {{ $.Chart.Name }}
+    chart: {{ $.Chart.Name }}-{{ $.Chart.Version }}
+    release: {{ $.Release.Name }}
+    heritage: {{ $.Release.Service }}
```

Comment:
This whole list of resources below here is pretty verbose and has a lot of the same data. I wonder if we could restructure it into a function or template so we could just feed it `[[resource, api_group, [verbs]], [resource, api_group, [verbs]]]` and maintain that in a yaml file or something.

Reply:
yes it has same verbs - we wanted to disable them based on what collectors are enabled or disabled

Reply:
Acutally, all the verbs are `["list", "watch"]` so if we could turn that long list into a function that generates the list, it would be much more succinct.

Reply:
our initial structure was combined one so with lots of resources included with common verbs now we have taken out a approach to enable/disable based on collectors